Mathlib/Lean/Meta/RefinedDiscrTree/Initialize.lean

```diff
@@ -0,0 +1,296 @@
+
+/--
+Structure for quickly initializing a lazy discrimination tree with a large number
+of elements using concurrent functions for generating entries.
+-/
+structure PreDiscrTree (α : Type) where
+  /-- Maps keys to index in tries array. -/
+  root : Std.HashMap Key Nat := {}
+  /-- Lazy entries for root of trie. -/
+  tries : Array (Array (LazyEntry × α)) := #[]
+  deriving Inhabited
+
+namespace PreDiscrTree
+
+@[specialize]
+private def modifyAt (d : PreDiscrTree α) (k : Key)
+    (f : Array (LazyEntry × α) → Array (LazyEntry × α)) : PreDiscrTree α :=
+  let { root, tries } := d
+  match root[k]? with
+  | none =>
+    { root := root.insert k tries.size, tries := tries.push (f #[]) }
+  | some i =>
+    { root, tries := tries.modify i f }
+
+/-- Add an entry to the pre-discrimination tree.-/
+def push (d : PreDiscrTree α) (k : Key) (e : LazyEntry × α) : PreDiscrTree α :=
+  d.modifyAt k (·.push e)
+
+/-- Convert a pre-discrimination tree to a `RefinedDiscrTree`. -/
+def toRefinedDiscrTree (d : PreDiscrTree α) : RefinedDiscrTree α :=
+  let { root, tries } := d
+  { root, tries := tries.map (.node #[] {} {}) }
+
+/-- Merge two discrimination trees. -/
+def append (x y : PreDiscrTree α) : PreDiscrTree α :=
+  let (x, y, f) :=
+        if x.root.size ≥ y.root.size then
+          (x, y, fun y x => x ++ y)
+        else
+          (y, x, fun x y => x ++ y)
+  let { root := yk, tries := ya } := y
+  yk.fold (init := x) fun d k yi => d.modifyAt k (f ya[yi]!)
+
+instance : Append (PreDiscrTree α) where
+  append := PreDiscrTree.append
+
+end PreDiscrTree
+
+
+/-- Information about a failed import. -/
+private structure ImportFailure where
+  /-- Module with constant that import failed on. -/
+  module  : Name
+  /-- Constant that import failed on. -/
+  const   : Name
+  /-- Exception that triggers error. -/
+  exception : Exception
+
+/-- Information generation from imported modules. -/
+private structure ImportErrorData where
+  errors : IO.Ref (Array ImportFailure)
+
+private def ImportErrorData.new : BaseIO ImportErrorData := do
+  return { errors := ← IO.mkRef #[] }
+/--
+Add the entries generated by `act name constInfo` to the `PreDiscrTree`.
+
+Note: It is expensive to create two new `IO.Ref`s for every `MetaM` operation,
+  so instead we reuse the same refs `mstate` and `cstate`. These are also used to
+  remember the cache, and the namegenerator across the operations.
+-/
+@[inline] private def addConstToPreDiscrTree
+    (cctx : Core.Context)
+    (env : Environment)
+    (modName : Name)
+    (data : ImportErrorData)
+    (mstate : IO.Ref Meta.State)
+    (cstate : IO.Ref Core.State)
+    (act : Name → ConstantInfo → MetaM (List (α × List (Key × LazyEntry))))
+    (tree : PreDiscrTree α) (name : Name) (constInfo : ConstantInfo) :
+    BaseIO (PreDiscrTree α) := do
+  if constInfo.isUnsafe then pure tree else
+  if LazyDiscrTree.blacklistInsertion env name then pure tree else
```

Comment:
style nit:
```suggestion
  if constInfo.isUnsafe then return tree
  if LazyDiscrTree.blacklistInsertion env name then return tree
```

Reply:
This code is performance critical since it gets run on most constants in the environment (when building the tree the first time). And the code doen't get compiled (instead it is run by the interpreter). When writing `if .. then return ..`, the compiler creates a separate function for the (implicit) else branch of the if statement, which is less efficient than a basic if-else.

Reply:
Could you add a comment in the code to this effect?